From 253a0fc5e2dfed8ad7bb98cb4320f6dc1407b6cd Mon Sep 17 00:00:00 2001
From: Sagar Cheluvegowda <quic_scheluve@quicinc.com>
Date: Mon, 10 Jun 2024 18:28:50 -0700
Subject: [PATCH 05/28] PENDING: net: stmmac: Add sa8797p support within ethqos
 driver

1.Add sa8797p support within ethqos driver by adding
corresponding compatible property and POR values.
2.Add support for EMAC version 6 devices.

Change-Id: If86702b1d725342f44e291c59e7d6747c0777adf
Signed-off-by: Sagar Cheluvegowda <quic_scheluve@quicinc.com>
(cherry picked from commit 276969829eb3fdaff1e31d386550b835f0e494cb)
---
 .../stmicro/stmmac/dwmac-qcom-ethqos.c        | 37 +++++++++++++++++++
 drivers/net/ethernet/stmicro/stmmac/hwif.c    |  4 +-
 include/linux/stmmac.h                        | 20 ++++++++--
 3 files changed, 57 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c b/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c
index ac7d6d3e205a1..82e9739c253b6 100644
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c
@@ -88,6 +88,7 @@ struct ethqos_emac_por {
 
 struct ethqos_emac_driver_data {
 	const struct ethqos_emac_por *rgmii_por;
+	struct dwxgmac_addrs dwxgmac_addrs;
 	unsigned int num_rgmii_por;
 	bool rgmii_config_loopback_en;
 	bool has_emac_ge_3;
@@ -95,6 +96,7 @@ struct ethqos_emac_driver_data {
 	const char *link_clk_name;
 	struct dwmac4_addrs dwmac4_addrs;
 	bool needs_sgmii_loopback;
+	bool has_xgmac;
 };
 
 struct qcom_ethqos {
@@ -277,6 +279,16 @@ static const struct ethqos_emac_por emac_v4_0_0_por[] = {
 	{ .offset = RGMII_IO_MACRO_CONFIG2,	.value = 0x00002060 },
 };
 
+static const struct ethqos_emac_por emac_v6_6_0_por[] = {
+	{ .offset = RGMII_IO_MACRO_CONFIG,	.value = 0x40c04c03 },
+	{ .offset = SDCC_HC_REG_DLL_CONFIG,	.value = 0xd642c },
+	{ .offset = SDCC_HC_REG_DDR_CONFIG,	.value = 0x80040868 },
+	{ .offset = SDCC_HC_REG_DLL_CONFIG2,	.value = 0xa001 },
+	{ .offset = SDCC_USR_CTL,		.value = 0x90106c0 },
+	{ .offset = RGMII_IO_MACRO_CONFIG2,	.value = 0x2200a0 },
+	{ .offset = SDCC_TEST_CTL, .value = 0x90106c0 },
+};
+
 static const struct ethqos_emac_driver_data emac_v4_0_0_data = {
 	.rgmii_por = emac_v4_0_0_por,
 	.num_rgmii_por = ARRAY_SIZE(emac_v4_0_0_por),
@@ -303,6 +315,25 @@ static const struct ethqos_emac_driver_data emac_v4_0_0_data = {
 	},
 };
 
+static const struct ethqos_emac_driver_data emac_v6_6_0_data = {
+	.rgmii_por = emac_v6_6_0_por,
+	.num_rgmii_por = ARRAY_SIZE(emac_v6_6_0_por),
+	.rgmii_config_loopback_en = false,
+	.dma_addr_width = 32,
+	.link_clk_name = "phyaux",
+	.has_xgmac = true,
+	.dwxgmac_addrs = {
+		.dma_even_chan_base  = 0x00008500,
+		.dma_odd_chan_base = 0x00008580,
+		.dma_chan_offset = 0x00001000,
+		.mtl_chan_base = 0x00008000,
+		.mtl_chan_offset =  0x00001000,
+		.timestamp_base = 0x00007000,
+		.pps_base = 0x00007080,
+		.pps_offset = 0x10,
+	},
+};
+
 static int ethqos_dll_configure(struct qcom_ethqos *ethqos)
 {
 	struct device *dev = &ethqos->pdev->dev;
@@ -778,6 +809,10 @@ static int qcom_ethqos_probe(struct platform_device *pdev)
 	if (ethqos->has_emac_ge_3)
 		plat_dat->dwmac4_addrs = &data->dwmac4_addrs;
 	plat_dat->pmt = true;
+	if (data->has_xgmac) {
+		plat_dat->core_type = DWMAC_CORE_XGMAC;
+		plat_dat->dwxgmac_addrs = &data->dwxgmac_addrs;
+	}
 	if (of_property_read_bool(np, "snps,tso"))
 		plat_dat->flags |= STMMAC_FLAG_TSO_EN;
 	if (of_device_is_compatible(np, "qcom,qcs404-ethqos"))
@@ -802,6 +837,8 @@ static const struct of_device_id qcom_ethqos_match[] = {
 	{ .compatible = "qcom,sa8775p-ethqos", .data = &emac_v4_0_0_data},
 	{ .compatible = "qcom,sc8280xp-ethqos", .data = &emac_v3_0_0_data},
 	{ .compatible = "qcom,sm8150-ethqos", .data = &emac_v2_1_0_data},
+	{ .compatible = "qcom,sa8620p-ethqos", .data = &emac_v4_0_0_data},
+	{ .compatible = "qcom,sa8797p-ethqos", .data = &emac_v6_6_0_data},
 	{ }
 };
 MODULE_DEVICE_TABLE(of, qcom_ethqos_match);
diff --git a/drivers/net/ethernet/stmicro/stmmac/hwif.c b/drivers/net/ethernet/stmicro/stmmac/hwif.c
index 265671170bf6e..9015f4fbcfe1e 100644
--- a/drivers/net/ethernet/stmicro/stmmac/hwif.c
+++ b/drivers/net/ethernet/stmicro/stmmac/hwif.c
@@ -316,6 +316,7 @@ stmmac_hwif_find(enum dwmac_core_type core_type, u8 snpsver, u8 dev_id)
 
 int stmmac_hwif_init(struct stmmac_priv *priv)
 {
+	const struct dwxgmac_addrs *dwxgmac_addrs = priv->plat->dwxgmac_addrs;
 	enum dwmac_core_type core_type = priv->plat->core_type;
 	const struct stmmac_hwif_entry *entry;
 	struct stmmac_version version;
@@ -369,7 +370,8 @@ int stmmac_hwif_init(struct stmmac_priv *priv)
 
 	priv->hw = mac;
 	priv->fpe_cfg.reg = entry->regs.fpe_reg;
-	priv->ptpaddr = priv->ioaddr + entry->regs.ptp_off;
+	priv->ptpaddr = priv->ioaddr +
+		(dwxgmac_addrs ? dwxgmac_addrs->timestamp_base : entry->regs.ptp_off);
 	priv->mmcaddr = priv->ioaddr + entry->regs.mmc_off;
 	memcpy(&priv->ptp_clock_ops, entry->ptp,
 	       sizeof(struct ptp_clock_info));
diff --git a/include/linux/stmmac.h b/include/linux/stmmac.h
index f1d393dfeb6b6..f7c7ba2028ac3 100644
--- a/include/linux/stmmac.h
+++ b/include/linux/stmmac.h
@@ -15,9 +15,9 @@
 #include <linux/platform_device.h>
 #include <linux/phylink.h>
 
-#define MTL_MAX_RX_QUEUES	8
-#define MTL_MAX_TX_QUEUES	8
-#define STMMAC_CH_MAX		8
+#define MTL_MAX_RX_QUEUES	12
+#define MTL_MAX_TX_QUEUES	12
+#define STMMAC_CH_MAX		12
 
 #define STMMAC_RX_COE_NONE	0
 #define STMMAC_RX_COE_TYPE1	1
@@ -195,6 +195,19 @@ enum dwmac_core_type {
 	DWMAC_CORE_XGMAC,
 };
 
+/* Addresses that may be customized by a platform */
+struct dwxgmac_addrs {
+	u32 dma_even_chan_base;
+	u32 dma_odd_chan_base;
+	u32 dma_chan_offset;
+	u32 mtl_chan_base;
+	u32 mtl_chan_offset;
+	u32 timestamp_base;
+	u32 pps_base;
+	u32 pps_offset;
+};
+
+#define STMMAC_FLAG_HAS_INTEGRATED_PCS		BIT(0)
 #define STMMAC_FLAG_SPH_DISABLE			BIT(1)
 #define STMMAC_FLAG_USE_PHY_WOL			BIT(2)
 #define STMMAC_FLAG_HAS_SUN8I			BIT(3)
@@ -358,6 +371,7 @@ struct plat_stmmacenet_data {
 	int msi_rx_base_vec;
 	int msi_tx_base_vec;
 	const struct dwmac4_addrs *dwmac4_addrs;
+	const struct dwxgmac_addrs *dwxgmac_addrs;
 	unsigned int flags;
 	struct stmmac_dma_cfg __dma_cfg;
 };
